### PR TITLE
feat(subscriptions): Insight Subscriptions

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

See https://github.com/PostHog/posthog/issues/9997

## Changes

* New "subscribe" option on Insight view opening a modal to manage subscriptions
* Addition of maildev for local testing of emails
* Background worker task to send all necessary emails and schedule follow up ones

Related to [Figma designs](https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=10306%3A42331)

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

* Some automated tests on the backend side of things
* Added a helpful email output-er that renders the to-be-sent email locally for checking it out visually
* Lots of manual UI testing

## Still TODO
- [x] Public unsubscribe page
- [x] Initial email to notify user they have been subscribed
- [x] Expand recurrence to allow picking weekday / day of month
- [x] Ensure unload is caught when making changes to a subscription
- [x] Publicly accessible exports (using some sort of token)
- [x] Logic tests
- [x] Pull out DB migration to other PR
- [x] Add message to Subscription sending
- [x] Update Exporter to style the exported image to match what is expected for the email

Depends on https://github.com/PostHog/posthog/pull/10292

### CODE REVIEW

File: frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    start_date: dayjs().hour(9).minute(0).second(0).toISOString(),
+    title: 'New Subscription',
+    target_type: 'email',
+}
+
+export interface InsightSubscriptionLogicProps {
+    id: number | 'new'
+    insightShortId: InsightShortId
+}
+export const insightSubscriptionLogic = kea<insightSubscriptionLogicType>([
```
Comment (pauldambra) [inline frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts:29]: flyby without proper review...

we should get tests around the logics before they (maybe possibly in the future) get more complicated or it is painful to backfill

Comment (benjackwhite) [reply frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts:29]: I love that one of my first comments on a PR of yours was - "maybe this should have tests" and now I'm getting the same treatment. You're keeping me an honest man 😉

Totally agreed. I've generally been quite lax on the frontend tests which i should try and rectify I think

Comment (pauldambra) [reply frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts:29]: spiderhogs-54f1bf6e8ebf31791e4c5c1813b19f2a

File: posthog/migrations/0241_subscription.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Generated by Django 3.2.13 on 2022-06-09 06:51
+
+from django.conf import settings
+import django.contrib.postgres.fields
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+class Migration(migrations.Migration):
```
Comment (macobo) [inline posthog/migrations/0241_subscription.py:9]: Note this needs to be in a separate PR - otherwise we are not able to roll this one back.

File: posthog/models/exported_asset.py [hunk 3]
```diff
@@ -54,3 +55,12 @@ def filename(self):
 
     def get_analytics_metadata(self):
         return {"export_format": self.export_format, "dashboard_id": self.dashboard_id, "insight_id": self.insight_id}
+
+    @property
+    def url(self):
+        return f"{settings.SITE_URL}/exports/{self.access_token}"
```
Comment (macobo) [inline posthog/models/exported_asset.py:61]: Problem: AFAIK SITE_URL might be unset - we don't set it by default in the chart for example.

Comment (benjackwhite) [reply posthog/models/exported_asset.py:61]: I think we do set it in the chart https://github.com/PostHog/charts-clickhouse/blob/d168b0f53f7540f00471dbc23c4fec8c0af2ca53/charts/posthog/templates/_helpers.tpl#L156
but its a good point that we should probably be checking more explicitly otherwise error message could be confusing

File: frontend/src/scenes/insights/Insight.tsx [hunk 4]
```diff
@@ -111,6 +113,14 @@ export function Insight({ insightId }: { insightId: InsightShortId | 'new' }): J
 
     const insightScene = (
         <div className="insights-page">
+            {insightId !== 'new' && (
+                <InsightSubscriptionsModal
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/scenes/insights/Insight.tsx:117]: Agreed. I have it behind the existing "export" feature flag so there are no buttons showing but probably makes sense to give it it's own dedicated flag

File: posthog/migrations/0241_subscription.py (hunk 2 above)
Comment (macobo) [reply posthog/migrations/0241_subscription.py:9]: The issue is that these PRs are not practically revertable (we'd get into a situation where deploys don't occur) so you need to be quite sure it's safe. 56 files isn't that :sweat_smile: 

I remembered this being documented in the handbook but couldn't find it. Here's a related incident though: https://github.com/PostHog/postmortem/blob/master/2022-02-18-rename-tags-column.md

File: posthog/models/exported_asset.py (hunk 3 above)
Comment (macobo) [reply posthog/models/exported_asset.py:61]: Note we do some stuff in ingestion that affects how e.g. webhooks do it. See https://github.com/PostHog/posthog/issues/9790 and linked PRs.

Comment (benjackwhite) [reply posthog/models/exported_asset.py:61]: Hmmmm question is should I use that dynamic value or the configured SITE_URL? If I would use it then the naming is a little off...

I wonder what was the motivation of using a different value to `SITE_URL`?

Comment (benjackwhite) [reply posthog/models/exported_asset.py:61]: Added a notice about SITE_URL to the subscription modal for now. Will create a follow up issue to discuss making SITE_URL configurable

**Review by mariusandra — COMMENTED** [review]
Managed to get through the frontend part, will continue with the backend now. Submitting some thoughts for consideration since it's a "one day PR" in reviewing time :D

- The "added you to a subscription" mail has a few bugs (broken img, :


- Something's wrong with these dropdowns:


- It also looks like I can basically select between midnight and 10am from what I do see in the dropdown. Intentional?

File: frontend/src/lib/components/InsightSubscription/InsightSubscriptionsModal.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import { InsightShortId, Realm, SubscriptionType } from '~/types'
+import preflightJson from '~/mocks/fixtures/_preflight.json'
+import { preflightLogic } from 'scenes/PreflightCheck/preflightLogic'
+import { insightSubscriptionLogic } from './insightSubscriptionLogic'
+import { insightSubscriptionsLogic } from './insightSubscriptionsLogic'
+
+export default {
+    title: 'Components/Subscription Modal',
+    component: InsightSubscriptionsModal,
+} as ComponentMeta<typeof InsightSubscriptionsModal>
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/lib/components/InsightSubscription/InsightSubscriptionsModal.stories.tsx:14]: It's all a bit hacky because the modal would override a bunch of styles. I could inline the content itself as they are already abstracted as separate components but then it wouldn't be that useful as a storybook...

File: frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    path(['lib', 'components', 'InsightSubscription', 'insightSubscriptionLogic']),
+    props({} as InsightSubscriptionLogicProps),
+    key(({ id, insightShortId }) => `${insightShortId}-${id ?? 'new'}`),
+    connect(({ insightShortId }: InsightSubscriptionLogicProps) => ({
+        values: [insightLogic({ dashboardItemId: insightShortId }), ['insight']],
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts:35]: Agreed

File: frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+export interface InsightSubscriptionLogicProps {
+    id: number | 'new'
+    insightShortId: InsightShortId
+}
+export const insightSubscriptionLogic = kea<insightSubscriptionLogicType>([
+    path(['lib', 'components', 'InsightSubscription', 'insightSubscriptionLogic']),
+    props({} as InsightSubscriptionLogicProps),
+    key(({ id, insightShortId }) => `${insightShortId}-${id ?? 'new'}`),
+    connect(({ insightShortId }: InsightSubscriptionLogicProps) => ({
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts:34]: Yup doesn't like it if its not there

File: frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                    target_type == 'email'
+                        ? !target_value
+                            ? 'At least one email is required'
+                            : !target_value.split(',').every((email) => isEmail(email))
+                            ? 'All emails must be valid'
+                            : undefined
+                        : undefined,
+            }),
+            submit: async (subscription) => {
+                subscription.insight = values.insight.id
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts:68]: Fixed it by just creating a new object

File: frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                let subscriptionId = props.id
+
+                if (props.id === 'new') {
+                    const newSub = await api.subscriptions.create(subscription)
+                    subscriptionId = newSub.id
+                } else {
+                    await api.subscriptions.update(props.id, subscription)
+                }
+                actions.loadSubscriptions()
+                actions.loadSubscription()
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/lib/components/InsightSubscription/insightSubscriptionLogic.ts:79]: Not sure what you mean by "add a breakpoint". Could you expand with a code example perhaps?

**Comment by benjackwhite** [comment]
In response to https://github.com/PostHog/posthog/pull/10197#pullrequestreview-1004172786
@mariusandra 
1. Img's in email doesn't work as it isn't implemented yet 😛 - see TODOs
2. Select dropdown is indeed a bug as the LemonSelect doesn't seem to keep to the window bounds. Might be a tricky fix...

File: frontend/src/lib/components/InsightSubscription/insightSubscriptionsLogic.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    loaders(({ values }) => ({
+        subscriptions: {
+            __default: [] as SubscriptionType[],
+            loadSubscriptions: async () => {
+                if (!values.insight.id) {
+                    return []
+                }
+
+                const response = await api.subscriptions.list(values.insight.id)
+                return response.results
```
[the remark this answers is not in the document]
Comment (mariusandra) [reply frontend/src/lib/components/InsightSubscription/insightSubscriptionsLogic.ts:35]: What can I add here to clarify it more? https://keajs.org/docs/core/listeners#breakpoints

**Review by mariusandra — COMMENTED** [review]
Finally got through the full set. Some more comments inline and a few below.

Overall, I think we could make the scheduling code a lot more robust. A good rule of thumb is to assume that any consecutive 15min can take anywhere between 0 and 60 minutes in reality due to various issues that are a part of live ("site down!"). Thus coding as code related to timing and subscriptions as defensively as possible will help. In practical terms, I left some thoughts on how to organise the next delivery date... but feel free to push back.

Some other points:

- After unsubscribing, I still had one active subscription:
image

- ... with nobody inside:
image

- I'd like to see the names / avatars of the users I subscribe this to, as I might now know everyone by their e-mail:
image

- If you open an old email, will you see the latest export png, or still the old ones?

File: frontend/src/lib/components/InsightSubscription/insightSubscriptionsLogic.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/lib/components/InsightSubscription/insightSubscriptionsLogic.ts:35]: Maybe adding search to the docs as I hadn't found that before 😅

File: posthog/models/subscription.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+def get_unsubscribe_token(subscription: Subscription, email: str) -> str:
+    encoded_jwt = jwt.encode(
+        {
+            "id": subscription.id,
+            "email": email,
+            "exp": datetime.now(tz=timezone.utc) + timedelta(days=UNSUBSCRIBE_TOKEN_EXP_DAYS),
+        },
+        settings.SECRET_KEY,
+        algorithm="HS256",
+    )
```
[the remark this answers is not in the document]
Comment (guidoiaquinti) [reply posthog/models/subscription.py:136]: I don't know much about JWT but I see the library does not natively expose a method argument for a salt. In the past I've always used it to add some pepper to the key keeping the scope limited by feature. Example:

```python
from itsdangerous import BadSignature, SignatureExpired, URLSafeTimedSerializer

def generate_email_verification_token(user_email):
    ts = URLSafeTimedSerializer(settings.SECRET_KEY)
    return ts.dumps(user_email, salt=EMAIL_VERIFICATION_SALT)

def check_email_verification_token(token):
    ts = URLSafeTimedSerializer(settings.SECRET_KEY)
    try:
        email = ts.loads(token, salt=EMAIL_VERIFICATION_SALT, max_age=86400)
    except (BadSignature, SignatureExpired):
        return None
    else:
        return email
 ```
 
 My understanding of JWT is not good enough to understand if this make sense here or not (likely not). Maybe @hazzadous might know more.

File: requirements.in [hunk 12]
```diff
@@ -50,6 +50,7 @@ pickleshare==0.7.5
 posthoganalytics==1.4.7
 protobuf==3.13.0
 psycopg2-binary==2.8.6
+pyjwt==2.1.0
```
Comment (guidoiaquinti) [inline requirements.in:53]: Why using `2.1.0` while `2.4.0` is available?

File: posthog/tasks/subscriptions.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+            try:
+                send_email_subscription_report(email, subscription, asset)
+            except Exception as e:
+                logger.error(e)
+                raise e
+    else:
+        raise NotImplementedError(f"{subscription.target_type} is not supported")
+
+    subscription.set_next_delivery_date(subscription.next_delivery_date)
+    subscription.save()
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply posthog/tasks/subscriptions.py:88]: Not sure how much a lock would help us here. The export itself is unlikely to take long but it may be that the number of tasks to be done takes so long that the next schedule comes along. In that case we rather need some sort of idempotency check to ensure we don't schedule something that is already scheduled. 

If we set the next delivery date on the schedule call instead that would solve this issue but then we lose the "retry-ability" side of things. That might not be so important though

File: requirements.in (hunk 12 above)
Comment (benjackwhite) [reply requirements.in:53]: I only hardcoded it to the existing version in requirements.txt (its a dependency of some other lib). I'll push it up

File: posthog/models/subscription.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (pauldambra) [reply posthog/models/subscription.py:136]: It looks like all JWTs this library generates (or at least from this encode method) are really JWS

https://github.com/jpadilla/pyjwt/blob/master/jwt/api_jwt.py#L64

So the secret is being used to sign the payload and isn't included in it.

(https://medium.facilelogin.com/jwt-jws-and-jwe-for-not-so-dummies-b63310d201a3 if you're not "familiar" with JWT vs JWS)

[the remark this answers is not in the document]
Comment (benjackwhite) [reply posthog/models/subscription.py]: One thing I forgot about (which I just added) is to scope it with an audience to indicate the token is clearly only for unsubscribing. It's not the same as salting though

[the remark this answers is not in the document]
Comment (pauldambra) [reply posthog/models/subscription.py]: Sidestep it completely...

Add a new secret that's only for signing these JWTs. Someone steals it they can't do anything except sign unsubscriptions and we can change it without impacting other features

**Comment by benjackwhite** [comment]
@marius LemonSelect bug is fixed here https://github.com/PostHog/posthog/pull/10278

File: posthog/models/subscription.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (benjackwhite) [reply posthog/models/subscription.py]: That would involve a fair amount of extra infra code which feels off to me. We could basically achieve the same though by hashing the SECRET_KEY to a different secret using "email:unsubscribe" as the salt?

All that said, I don't see any immediate danger of using the SECRET_KEY itself...

**Comment by benjackwhite** [comment]
@mariusandra 
* "After unsubscribing, I still had one active subscription:" - ✅ fixed
* "I'd like to see the names / avatars of the users I subscribe this to, as I might now know everyone by their e-mail:
  - Not sure if we want this. As nice as it looks it may be confusing to users typing in external emails? @clarkus what do you think? There is a nicer design in place once we update away from the Antd Select but opted not to jump into that yet (this PR is already big enough)
* "If you open an old email, will you see the latest export png, or still the old ones?" - You would see the last one which is what I would for sure expect.

**Comment by clarkus** [comment]
> Not sure if we want this. As nice as it looks it may be confusing to users typing in external emails? @clarkus what do you think? There is a nicer design in place once we update away from the Antd Select but opted not to jump into that yet (this PR is already big enough)

I think it depends on what we know about the users. In some cases, if the email corresponds to a PostHog account with more details, we could show that here. We do support sending emails to folks without accounts, so with that in mind we'll have scenarios where we don't have any other info than the email address. 

It also depends on how the multiple select control works. If a user can edit an entry, then it could be awkward if that entry has changed format from email to real name. Our current multiple select doesn't allow for this (you remove and re-add the person instead of editing a specific entry), but if that changes, it'd impact what we decide to show here. 

TL;DR - if we know the name of the user and can show it, then that's a good thing. It'd also help differentiate folks with accounts from folks without accounts / accounts with no profile information. For entries that change format to show names, it'd be good to show the email address via some secondary interaction. For example, if you hover over the name you see the corresponding email value.

File: posthog/tasks/subscriptions.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (benjackwhite) [reply posthog/tasks/subscriptions.py:102]: I changed the logic slightly so it triggers 5 minutes before and allows a buffer. Given the subs are checked every hour this should work. My vote is to get this merged as is and then test it a bit, and decide whether we need more controls

**Review by mariusandra — APPROVED** [review]
Still found this:



And a test is failing.

However here's a ✅  so you can merge in this when the other two PRs get in

File: frontend/src/lib/components/InsightSubscription/utils.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    '-1': { label: 'last' },
+}
+
+export const timeOptions: LemonSelectOptions = range(0, 24).reduce(
+    (acc, x) => ({
+        ...acc,
+        [String(x)]: { label: `${String(x).padStart(2, '0')}:00` },
+    }),
+    {}
+)
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/lib/components/InsightSubscription/utils.ts:47]: Thats a nice idea

[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/lib/components/InsightSubscription/utils.ts:47]: I might catch this in a different PR as this one is exhausting 😅

**Comment by benjackwhite** [comment]
@mariusandra that bug you found seems to have come in since I added the `breakpoint()` here https://github.com/PostHog/posthog/pull/10197/files#diff-511d3da7b12653efe350390dbf85c06beb321056e318f9bbe4cc4362abc829eeR90

Still don't fully understand them...
